src/clients/transmission.py

```diff
@@ -0,0 +1,131 @@
+      }
+    else:
+      raise TorrentClientError("Client returned unexpected response")
+
+  def inject_torrent(self, source_torrent_infohash, new_torrent_filepath, save_path_override=None):
+    source_torrent_info = self.get_torrent_info(source_torrent_infohash)
+
+    if not source_torrent_info["complete"]:
+      raise TorrentClientError("Cannot inject a torrent that is not complete")
+
+    new_torrent_infohash = calculate_infohash(get_bencoded_data(new_torrent_filepath)).lower()
+    new_torrent_already_exists = self.__does_torrent_exist_in_client(new_torrent_infohash)
+    if new_torrent_already_exists:
+      raise TorrentExistsInClientError(f"New torrent already exists in client ({new_torrent_infohash})")
+
+    self.__wrap_request("torrent-add", arguments={
+      "download-dir": save_path_override if save_path_override else source_torrent_info["save_path"],
+      "metainfo": base64.b64encode(open(new_torrent_filepath, "rb").read()).decode("utf-8"),
+      "labels": source_torrent_info['label']
+    })
+
+    return new_torrent_infohash
+
+  def __authenticate(self):
+    try:
+      # This method specifically does not use the __wrap_request method
+      # because we want to avoid an infinite loop of re-authenticating
+      response = requests.post(self._base_url, auth=self._basic_auth)
+      if response.status_code == HTTPStatus.CONFLICT:
+        self._transmission_session_id = response.headers[self.X_TRANSMISSION_SESSION_ID]
+      else:
+        response.raise_for_status()
+    except requests.RequestException as e:
+      raise TorrentClientAuthenticationError(f"TransmissionBt login failed: {e}")
+
+    if not self._transmission_session_id:
+      raise TorrentClientAuthenticationError("TransmissionBt login failed: Invalid username or password")
+
+  def __wrap_request(self, method, arguments, files=None):
+    try:
+      return self.__request(method, arguments, files)
+    except TorrentClientAuthenticationError:
+      self.__authenticate()
+      return self.__request(method, arguments, files)
+
+  def __request(self, method, arguments=None, files=None):
+    try:
+      response = requests.post(
+        self._base_url,
+        auth=self._basic_auth,
+        headers=CaseInsensitiveDict({self.X_TRANSMISSION_SESSION_ID: self._transmission_session_id}),
+        json={
+          "method": method,
+          "arguments": arguments
+        },
+        files=files,
+      )
+
+      response.raise_for_status()
+
+      return response.text
+    except requests.RequestException as e:
+      if e.response.status_code == 401:
+        print(e.response.text)
+        raise TorrentClientAuthenticationError("Failed to authenticate with TransmissionBt")
+
+      raise TorrentClientError(f"TransmissionBt request to '{self._base_url}' for method '{method}' failed: {e}")
+
+  def __does_torrent_exist_in_client(self, infohash):
+    try:
+      return bool(self.get_torrent_info(infohash))
```

Comment:
Just to confirm, If I'm not mistaken, based on this either returning a crafted dict or raising, this would always return true or throw->catch->return False?

Not that there is anything wrong here, but I'm just making sure that is the intent.

Also, this doesn't leave any error handling for the json.load in the even of transient errors during JSON deserialization, which would entirely be possible.

Reply:
Yes, this was copied from your qbittorrent implementation.

Reply:
I'm not the maintainer here, I'm just a contributor. @moleculekayak did the implementations of the clients.

I'm not intimately familiar with them other than the high-level layout and flow of the Deluge implementation.

I'm just merely pointing out what I noticed, if it applies to the qBittorrent (or deluge) code as well, probably should add handling for that too I guess.

Reply:
Oh, my bad. Didn't notice you weren't the maintainer :/.

I do believe the error handling here is okay, since we only want to know if the torrent exists, which is only true when the function `get_torrent_info` returns anything. When torrent is not found, it simply throws error, which is caught here and False is returned.

Reply:
Sorry, I think my review comment may not give you the idea I'm trying to convey in regards to the error handling.

If you go up to where get_torrent_info() is def'd 

https://github.com/moleculekayak/fertilizer/pull/26/files#diff-53b9f9adf3f6693f0d577a1324ef1af94cbdf5938ec17d3d6f35899d64b14299R46

json.loads when fed malformed or unexpected data can raise JSONDecodeError. The scoped except will only catch `TorrentClientError` - so this is an unhandled exception that is not that unlikely, and very easy to account for.

Reply:
Oh, I see. I will update the code to catch it and map it to TorrentClientError, like in deluge. Thanks!